File: lib/mock/asset/tools/ml2fsm.py
from xml.etree.ElementTree import ElementTree

groupSep  = '.'
actionSep = ':'

ML_NS = '{http://graphml.graphdrawing.org/xmlns}'
Y_NS  = '{http://www.yworks.com/xml/graphml}'

FSM_NODE_NAME_ANY = 'ANY'

allNodes = {}
allConnections = {}


class FSMParseException( Exception ):
	pass

def getTag( node ):
	tag = node.tag
	index = tag.find( '}' )
	if index != -1:
		return tag[ index+1: ]
	else:
		return tag

class FSMConnection(object):
	def loadFromXML( self, xmlNode ):
		self.id = xmlNode.attrib['id']
		label = xmlNode.find( './/'+ Y_NS +'EdgeLabel' )
		if label is not None:
			self.value = label.text.strip()
			if not self.value:
				self.values = False
		else:
			self.value = False
		self.nodeIdFrom = xmlNode.attrib['source']
		self.nodeIdTo   = xmlNode.attrib['target']
		allConnections[ self.id ] = self

	def findNodes( self ):
		self.nodeFrom = allNodes[ self.nodeIdFrom ]
		self.nodeTo   = allNodes[ self.nodeIdTo ]

class FSMNode(object):
	def __init__( self ):
		self.value = None
		self.fullname = None
		self.funcname = None
		self.connections = []
		self.parent = None
		self.id = ''
		self.name = ''
		self.jump = {}
		self.next = {}

	def isAny( self ):
		return self.name == FSM_NODE_NAME_ANY

	def loadFromXML( self, xmlNode ):
		if xmlNode.tag == ML_NS + 'graphml':
			self.id = 'root'
			self.name = ''
		else:
			self.id = xmlNode.attrib['id']
			label = xmlNode.find( './/'+ Y_NS+'NodeLabel' )
			self.name = label.text		
		allNodes[ self.id ] = self

	def getType( self ):
		return 'state'

	def isGroup( self ):
		return False

	def getName( self ):
		return self.name

	def getFullName( self ):
		if not self.fullname:
			if self.parent:
				pn = self.parent.getFullName()
				if pn:
					self.fullname = pn + groupSep + self.name
				else:
					self.fullname = self.name
			else:
				self.fullname = False
		return self.fullname

	def getFuncName( self ):
		if not self.funcname:
			self.funcname = '_FSM_' + str.replace( self.getFullName(), '.' , '_' )
		return self.funcname


	def __repr__( self ):
		return str(self.getFullName())


class FSMNodeGroup( FSMNode):
	def __init__( self ):
		super( FSMNodeGroup, self ).__init__()
		self.children = []

	def loadFromXML( self, xmlNode ):
		super( FSMNodeGroup, self ).loadFromXML( xmlNode )
		graphNode = xmlNode.find( ML_NS+'graph' )
		self.loadChildrenFromXML( graphNode )

	def isGroup( self ):
		return True

	def getType( self ):
		return 'group'

	def loadChildrenFromXML( self, xmlNode ):
		for childXml in xmlNode:
			tag = getTag( childXml )
			if tag == 'node':
				if childXml.attrib.get( 'yfiles.foldertype' ) == 'group':
					child = FSMNodeGroup()
				else:
					child = FSMNode()
				child.parent = self
				child.loadFromXML( childXml )
				self.children.append( child )

			elif tag == 'edge':
				conn = FSMConnection()
				conn.loadFromXML( childXml )
				self.connections.append( conn )

	def findChild( self, name ):
		for child in self.children:
			if child.name == name:
				return child
		return False

	def getAllChildren( self ):
		result = self.children[:]
		for c in self.children:
			if c.isGroup():
				result += c.getAllChildren()
		return result

class FSMParser():
	def __init__( self ):
		self.output = ''

	def parse( self, fileName ):
		global allNodes
		global allConnections
		allNodes = {}
		allConnections = {}

		tree     = ElementTree()
		rootNode = tree.parse( fileName )
		root     = FSMNodeGroup()
		root.loadFromXML( rootNode )

		self.root = root

		for conn in allConnections.values():
			conn.findNodes()
		##############-
		#generate jump table
		for e in allConnections.values():
			nodeFrom   = e.nodeFrom
			nodeTo     = e.nodeTo
			msg        = e.value
			jump       = nodeFrom.jump
			isGroup    = False
			if nodeTo.isGroup() :
				isGroup = True
				startState = nodeTo.findChild( 'start' )
				if startState :
					nodeTo = startState
				else:
					raise FSMParseException( "no 'start' state for group:" + nodeTo.getFullName() )

			if msg and msg != '' :
				jump[msg]=nodeTo

			else:
				if nodeFrom.isAny():
					raise FSMParseException( "cannot jump from ANY state without msg/condition: " + nodeTo.getFullName() )

				#validate group
				if nodeFrom.isGroup() :
					stopState = nodeFrom.findChild( 'stop' )
					if stopState :
						nodeFrom=stopState					
					else:
						raise FSMParseException( "no 'stop' state for group:" + nodeFrom.getFullName() )
				
				next = nodeFrom.next
				if not next :
					next = {}
					nodeFrom.next = next
				
				next[ nodeTo ] = isGroup and 'group' or 'node'
		
		####
		for node in allNodes.values():
			self.updateJump( node )

		return self.generateCode()
	
	def generateJumpTarget( self, nodeFrom, nodeTo ):
		if nodeTo.isAny():
			raise FSMParseException( "cannot jump into ANY state : <-" + nodeFrom.getFullName() )

		if nodeFrom.parent == nodeTo.parent :
			return ( '"%s"' % nodeTo.getFullName() )

		exits  = ""
		enters = ""
		node = nodeFrom.parent
		while True: #find group crossing path
			found = False
			enters = ""
			node1 = nodeTo.parent
			while node1 != self.root:
				if node1 == node :
					found = True
					break 
				enters = ( '"%s",' % (node1.getFuncName() + '__jumpin' ) ) + enters
				node1 = node1.parent				
			
			if found :
				break
			if node == self.root :
				break 

			exits = exits + ( '"%s",' % ( node.getFuncName() + '__jumpout') )
			node = node.parent	
		
		output = exits + enters
		#format: [ list of func needed nodeTo be called ] + 'next state name'
		return '{ %s"%s" }' % ( output, nodeTo.getFullName() )
	
	#overwrite according nodeTo parent-level-priority
	def updateJump( self, node, src = None ):
		if not node or node == self.root : return 
		if src :
			jump0 = src.jump
			jump  = node.jump
			for msg, target in jump.items():
				jump0[ msg ] = target
		else:
			src=node
		
		return self.updateJump( node.parent, src )

	def processAnyJump( self, anyNode ):
		for msg, target in anyNode.jump.items():
			if target.isAny(): continue
			sources = anyNode.parent.getAllChildren()
			for n in sources:
				if n == self.root: continue
				if n.isAny(): continue
				if n == target: continue
				if msg in n.jump:
					#warn?
					print( 'warning: multiple jump target in "%s" for msg "%s"' % ( n.getFullName(), msg ) )
				else:
					n.jump[ msg ] = target

	def generateCode( self ):
		self.output = ''
	
		def writef( pattern, *args ):
			self.output = self.output + ( pattern % args )
		
		def write(a):
			self.output = self.output+a

		#data code(jumptable) generation
		# file=io.open(fnout,'w')
		# if not file : 
		# 	print("cannot open file nodeTo write")
		# 	os.exit(1)
		# 
		write( 'return ' )
		write( '(function()\n' )
		write( 'nodelist={' )
		write( '\n' )
		for id, n in allNodes.items():
			if n == self.root: continue
			if n.isAny() :continue
			writef( '["%s"]={name="%s",localName="%s",id="%s",type="%s"};',n.getFullName(),n.getFullName(),n.getName(),n.getFuncName(),n.getType() )
			write( '\n' )
		
		write( '};' )
		write( '\n' )

		#ANY jump
		write( '-----------\n' )
		for id, n in allNodes.items():
			if n == self.root: continue
			if not n.isAny() :continue
			self.processAnyJump( n )

		for id, n in allNodes.items():
			if n == self.root: continue
			if n.isAny() :continue
			write( '-----------\n' )
			if len( n.jump ) > 0:
				writef( 'nodelist["%s"].jump={\n',n.getFullName() )
				for msg, target in n.jump.items():
					jumpto = target.getFullName()
					writef( '\t["%s"]=%s;\n', msg, self.generateJumpTarget( n, target ) )
				
				write( '}\n' )
			else:
				writef( 'nodelist["%s"].jump=false\n',n.getFullName() )
			
			if len( n.next ) > 0: #non-conditional jump
				writef( 'nodelist["%s"].next={\n', n.getFullName() )
				count = 0
				for target,targetType in n.next.items():
				# target=n.next
					jumpto = target.getFullName()
					targetName = target.getName()
					#add a symbol for better distinction
					if targetType == 'group' :
						targetName = target.parent.getName()
					writef('["$%s"]=%s;\n',targetName, self.generateJumpTarget( n, target ) ) 
					count += 1
				
				if count == 1 : #only one, give it a 'True' transition
					target = list(n.next.keys())[0]
					jumpto = target.getFullName()
					writef( '[true]=%s;\n', self.generateJumpTarget( n, target ) )
				
				writef( '}\n' )
		
		write( '-----------\n' )
		write( 'return nodelist\n' )
		write( 'end) ()\n' )
		write( '\n' )
		return self.output

def convertGraphMLToFSM( input, output ):
	parser = FSMParser()
	result = parser.parse( input )
	with open( output, 'w' ) as fp:
		fp.write( result )
	return True

if __name__ == '__main__':
	convertGraphMLToFSM( 'test.fsm.graphml', 'fsm_test_output.lua' )

File: lib/mock/asset/tools/ml2fsm.lua
#!/usr/local/bin/lua

local input, output=...
if not input or not output then
	print(string.format('usage: ml2fsm.lua <input_file.graphml> <output_file.lua>'))
	return
end

lxp = require 'lxp'

local match = string.match
function string.trim(s)
	return match(s,'^()%s*$') and '' or match(s,'^%s*(.*%S)')
end

local function convertGraphML(fn)
	local callbacks={}
	
	local objects={}

	local root={value='nodes',type='group'}
	local edges={}

	local nodeStack={}
	local pointer=0
	local currentNode
	local prevNode

	local groupSep='.'
	local actionSep=':'

	local function getFullName(n)
		local res=n.value or ''
		while true do
			local p=n.parent
			if not p or p==root then return res end
			res=p.value..groupSep..res
			n=p
		end
	end

	local function pushNode(n)
		pointer=pointer+1
		nodeStack[pointer]=n
		
		local parent=currentNode or root
		local children=parent.children
		if not children then children={} parent.children=children end
		children[n]=true
		n.parent=parent

		currentNode=n
	end

	local function popNode()
		if pointer>0 then
			pointer=pointer-1
			currentNode=nodeStack[pointer]
		else
			currentNode=nil
		end
	end

	local currentGroup
	
	local function getObject(id)
		local o=objects[id]
		if not o then o={id=id} objects[id]=o end
		return o
	end

	local function CharacterData(parser,string)
		if not currentNode.value then currentNode.value=string:trim() end
	end

	function callbacks.StartElement(parser,name,attrs)
		callbacks.CharacterData=false
		if name=='edge' then			
			local n={
				type='edge',
				id=attrs.id,
				from=getObject(attrs.source),
				to=getObject(attrs.target),
			}
			edges[attrs.id]=n
			pushNode(n)
		elseif name=='node' then
			local n=getObject(attrs.id)
			n.jump={}
			if attrs['yfiles.foldertype']=='group' then
				n.type='group'
			else
				n.type='state'
			end
			pushNode(n)
		elseif name=='y:EdgeLabel' or name=='y:NodeLabel' then
			if currentNode and not currentNode.value then
				callbacks.CharacterData=CharacterData
			end
		end
	end

	function callbacks.EndElement(parser,name)
		if name=='node' then
			-- callbacks.CharacterData=false			
			currentNode.fullname=getFullName(currentNode)
			currentNode.funcname='_FSM_'..string.gsub(currentNode.fullname,'%.','_')
			popNode()
		-- elseif name=='y:EdgeLabel' or name=='y:NodeLabel' then
		-- 	callbacks.CharacterData=false
		end
	end

	callbacks.CharacterData=false

	local file=io.open(fn,'r')
	if not file then 
		print("file not found:",fn)
		os.exit(1)		
	end	

	local p=lxp.new(callbacks)

	for line in file:lines() do
		p:parse(line)
	end

	file:close()


	local function findchild(group,name)
		local children=group.children
		if not children then return nil end
		for c in pairs(children) do
			if c.value==name then return c end
		end
		return nil
	end
	-----------------------------
	--generate jump table
	for i,e in pairs(edges) do
		local from=e.from
		local to=e.to
		local msg=e.value

		local jump=from.jump
		local isGroup=false
		if to.type=='group' then
			isGroup=true
			local startState=findchild(to,'start')
			if startState then
				to=startState
			else
				print("no 'start' state for group:",to.fullname)
				return error()
			end
		end

		if msg and msg~='' then
			jump[msg]=to
		else
			--validate group
			if from.type=='group' then
				local stopState=findchild(from,'stop')
				if stopState then
					from=stopState					
				else
					print("no 'stop' state for group:",from.fullname)
					return error()
				end
			end
			
			local __next=from.__next
			if not __next then
				__next={}
				from.__next=__next
			end
			__next[to]=isGroup and 'group' or 'node'
		end

	end

	local function generateJumpTarget( from, to )
		if from.parent == to.parent then
			return string.format( '%q', to.fullname )
		end
		local exits  = ""
		local enters = ""
		local node = from.parent
		while true do --find group crossing path
			local found = false
			enters = ""
			local node1 = to.parent
			while node1 ~= root do
				if node1 == node then	found = true break end
				enters = string.format( '%q,', node1.funcname .. '__jumpin' ) .. enters
				node1 = node1.parent				
			end
			if found then
				break
			end
			if node == root then break end
			exits = exits .. string.format( '%q,', node.funcname .. '__jumpout' )
			node = node.parent	
		end
		local output = exits .. enters
		--format: [ list of func needed to be called ] .. 'next state name'
		return string.format( "{ %s%q }", output, to.fullname)		
	end

	--overwrite according to parent-level-priority
	local function updateJump(node,src)
		if not node or node==root then return end
		if src then
			local jump0=src.jump
			local jump=node.jump
			for msg,target in pairs(jump) do
				jump0[msg]=target
			end
		else
			src=node
		end
		return updateJump(node.parent,src)
	end

	for i,o in pairs(objects) do
		updateJump(o)
	end

	local output=''
	local function writef(...)
		output=output..(string.format(...))
	end

	local function write(a)
		output=output..a
	end

	--data code(jumptable) generation
	-- file=io.open(fnout,'w')
	-- if not file then 
	-- 	print("cannot open file to write")
	-- 	os.exit(1)
	-- end
	write('(function()\n')
	write('local nodelist={')
	write('\n')
	for id,n in pairs(objects) do		
		writef(string.format('[%q]={name=%q,localName=%q,id=%q,type=%q};',n.fullname,n.fullname,n.value,n.funcname,n.type))
		write('\n')
	end
	write('};')
	write('\n')


	for id,n in pairs(objects) do
		write('-----------\n')
		if n.jump and next(n.jump) then
			writef('nodelist[%q].jump={\n',n.fullname)
			for msg, target in pairs(n.jump) do
				local jumpto = target.fullname
				writef( '\t[%q]=%s;\n', msg, generateJumpTarget( n, target ) )
			end
			write('}\n')
		else
			writef('nodelist[%q].jump=false\n',n.fullname)
		end

		if n.__next then
			writef('nodelist[%q].next={\n', n.fullname )
			local count=0
			for target,targetType in pairs(n.__next) do
			-- local target=n.__next
				local jumpto=target.fullname
				local targetName=target.value
				--add a symbol for better distinction
				if targetType=='group' then targetName=target.parent.value end
				writef('["$%s"]=%s;\n',targetName, generateJumpTarget( n, target ) ) 
				count=count+1
			end
			
			if count==1 then --only one, give it a 'true' transition
				local target=next(n.__next)
				local jumpto=target.fullname
				writef('[true]=%s;\n', generateJumpTarget( n, target ) )
			end

			writef('}\n')
		end
	end

	write('return nodelist\n')
	write('end) ()\n')

	write('\n')

	return output
end


local function stripExt(p)
		 p=string.gsub(p,'%..*$','')
		 return p
end

local function extractExt(p)
	return string.gsub(p,'.*%.','')
end


function convertSingle( input, output )
	local result = convertGraphML( input )
	local outfile=io.open(output,'w')
	outfile:write( 'return ')
	outfile:write( result )
	outfile:close()	
	return true
end

convertSingle( input, output )
